torch/onnx.py

```diff
@@ -23,35 +44,60 @@ def export(model, args, f, export_params=True, kwargs=None, verbose=False):
+            will become inputs of the exported model, in the order they
+            occur in args.  If args is a Variable, this is equivalent
+            to having called it with a 1-ary tuple of that Variable.
+            (Note: passing keyword arguments to the model is not currently
+            supported.  Give us a shout if you need it.)
         f: a file-like object (has to implement fileno that returns a file descriptor)
             or a string containing a file name.  A binary Protobuf will be written
             to this file.
         export_params (bool, default True): if specified, all parameters will
-            be exported.  Set this to False if you are exporting an
-            untrained model.
-        kwargs (dict, optional): keyword inputs to the model.
+            be exported.  Set this to False if you want to export an untrained model.
+            In this case, the exported model will first take all of its parameters
+            as arguments, the ordering as specified by ``model.state_dict().values()``
+        verbose (bool, default False): if specified, we will print out a debug
+            description of the trace being exported.
+        training (bool, default False): export the model in training mode.  At
+            the moment, ONNX is oriented towards exporting models for inference
+            only, so you will generally not need to set this to True.
     """
-    _export(model, args, f, export_params, kwargs, verbose)
+    _export(model, args, f, export_params, verbose, training)
 
 
-def _export(model, args, f, export_params=True, kwargs=None, verbose=False):
+def _export(model, args, f, export_params=True, verbose=False, training=False):
     # Special case for common case of passing a single Variable
     if isinstance(args, torch.autograd.Variable):
         args = (args, )
-    if not kwargs:
-        kwargs = {}
-    trace, torch_out = torch.jit.record_trace(model, *args, **kwargs)
+    # Look at the state_dict *prior* to running the model, as this
+    # accurately captures what inputs we actually passed to the model.
+    # If we run it afterwards, a buggy forward pass could have
+    # added/deleted parameters, changing the structure of the state_dict.
+    #
+    # Note that it's possible the actual /data/ may change by the time we
+    # actually export the model, because we're not cloning the parameters
+    # here.  This shouldn't happen, because we've turned off training
+    # mode and networks in inference mode should be purely functional.
+    # But it's user code; anything can happen!
+    params = None
+    if export_params:
+        # NB: OrderedDict values is not actually a list, but trace.export is
+        # not duck-typed and expects an actual list.
+        params = list(model.state_dict().values())
+    # It's important to run the model in inference mode when exporting;
+    # otherwise internal buffers may get updated, dropout gets applied, etc.
```

Comment:
this code doesn't hard-code inference mode. It's only `False` by default, but can be overriden